File: src/cse/ele/EleTuple.java
package cse.ele;

import java.util.Arrays;

/**
 * EleTuple is a container for several elements.

 */
public class EleTuple extends EleValueOrTuple {
    private final EleValueOrTuple[] value;

    /**
     * Create a tuple element with tuple label.
     */
    public EleTuple(EleValueOrTuple[] value) {
        super("tuple");
        this.value = value;
    }

    /**
     * Get the tuples inside the element.
     */
    public EleValueOrTuple[] getValue() {
        return value;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        EleTuple that = (EleTuple) o;
        return Arrays.equals(value, that.value);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(value);
    }

    @Override
    public String toString() {
        return Arrays.toString(value);
    }
}

File: reCreate/main/CSE/elements/EleTuple.py
from .EleValueOrTuple import EleValueOrTuple

class EleTuple(EleValueOrTuple):
    def __init__(self, value):
        super().__init__("tuple")
        self.value = value

    def getValue(self):
        return self.value

    def equals(self, o):
        if self is o:
            return True
        if o is None or self.__class__ != o.__class__:
            return False
        return self.value == o.value

    def hashCode(self):
        return hash(tuple(self.value))

    def toString(self):
        return str(self.value)


